### PR TITLE
🐳 Docker (Compose) support for GAwesomeBot

### PR DESCRIPTION
**Pull request by probwebdev**
**GawesomeBot Docker 101:**
0. Configure bot itself for dev or prod environment:
    - Copy .env.example to .env. Change settings if you wish. DOCKER_REPO is your public or private
    - MongoDB url should be like that mongodb://**db**:27017/gawesomebot?poolSize=10. If you want to use native mongodb instance, than docker-compose.yml files should be altered.
1. Run dev environment:
    - **docker-compose up** - this will run dev environment. Docker Compose will create a docker volume for mongodb storage. Bot sources will be **binded** into image except node_modules so you can alter code on the fly as you're working without docker. No need to rebuild image every time.
    1.1 Debug mode:
       - **docker-compose -f docker-compose.yml -f docker-compose.override.yml -f docker-compose.debug.yml**
2. Run _prod_ environment(need to discuss this one):
    2.1 Build GAwesomeBot image:
        - **docker build -t DOCKER_REPO/gawesomebot:_<insert your tag>_ .** - will build your image, also you can add additional tags if you want.
        - docker login
        - docker push DOCKER_REPO/gawesomebot
    2.2 Run in prod environment:
        - **docker-compose -f docker-compose.yml -f docker-compose.prod.yml up** - this will pull mongodb and gawesomebot images from docker hub. Mongo will use binded directory from your host that you specified during configuration step. **config.js**, **auth.js**, **config.json** and **logs** are binded directories from your host.
3. Travis:
    - Add DOCKER_USERNAME and DOCKER_PASSWORD environment variables
    - Currently travis will build and push docker image to the repo if travis build succeeded and current branch is master.

Discussion:
- I'm not sure about whole prod thing. Where and how we should store db data? Bind dir or Docker Volume? And other questions, too much of them.
- Travis could be improved. You can set tracked branches from which docker should build fresh images. Tags could be automated instead of just **:latest**
- Because i used shared docker-compose.yml files gui applications may be not aware of the docker-compose.override.yml. At some point we need to decide if we need shared configurations or copy-pasted.

Docker is just a layer, no need to use it at all. It's just an option.
Sorry if I forgot something. We can discuss anything here or on the bot server.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#DB settings
+DB_VER=3.4
```
Comment (vladfrangu) [inline .env.example]: Since this is pointed at indev, MongoDB should be `3.6`

Comment (probwebdev) [reply .env.example]: No issue here. I was using 3.6 at first but decided to downgrade to 3.4 *just-in-case* :)

  And it's just an example file, you can bump it during your environment configuration.

**Comment by Gilbert142** [comment]
Let us know when this PR is ready to be merged

**Comment by Gilbert142** [comment]
Fix the code checks and I'll merge

**Comment by probwebdev** [comment]
Gilbert, I will return to this PR on Saturday and Sunday. Want to change db storage method and additionally will fix code checks.

**Comment by TTtie** [comment]
fat ^^^

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,20 @@
+FROM node:carbon-alpine
```
[the remark this answers is not in the document]
Comment (KingDGrizzle) [reply Dockerfile]: Since when can you review!?

[the remark this answers is not in the document]
Comment (TTtie) [reply Dockerfile]: @KingDGrizzle the codacy settings thingy

**Comment by probwebdev** [comment]
Gilbert, seems that you resolved an issue with hadolint, thanks. MAINTAINER is deprecated so we should use LABEL instead. And I found out that hadolint fixed it. How did you resolved it exactly for Codacy?
Considering db storage. In the end I decided to go with external docker volume for mongodb storage. I think it's more persistent and you can use it for multiple containers. Anyway it could be altered with no hassle.

**Comment by Gilbert142** [comment]
@rgiyf The DB storage method is fine with me!

I fixed it by ignoring the issue on Codacy, so it doesn't count the PR as negative anymore.

**Comment by probwebdev** [comment]
Rewrite is required. The main blocker is compatibility issue between canvas-prebuilt package and alpine docker image. Additionally dockerfile and docker-compose files are poorly written.

**Pull request closed without merging**